### PR TITLE
[Carbon.MassTransit] Quorum Queues introduced

### PR DESCRIPTION
**Pull request by keremaytac**
You can create quorum queues, or migrate your classic queues to quorum queues automatically.

Beware of that, during the migration data loss may occur.

### CODE REVIEW

**Review by osman-koc — APPROVED** [review]

File: Carbon.MassTransit/IServiceCollectionConfiguratorExtensions.cs [hunk 1]
```diff
@@ -212,13 +212,15 @@ public static void AddAsyncRequestResponsePatternForRequestor<T>(this IServiceCo
                 {
                     busFactoryConfig.ReceiveEndpoint("request-starter-state", e =>
                     {
+                        e.AddAsHighAvailableQueue(configuration);
```
Comment (616e696c) [inline Carbon.MassTransit/IServiceCollectionConfiguratorExtensions.cs:215]: Why it's enabled by default on there? Shouldn't we also enable quorums on usual "AddRabbitMqBus" methods?

File: Carbon.MassTransit/IServiceCollectionConfiguratorExtensions.cs [hunk 2]
```diff
@@ -279,6 +282,27 @@ public static void AddAsyncRequestResponsePatternForResponder<T>(this IServiceCo
...
 
+        /// <summary>
+        /// ReceiveEndpoint queue will be declared as a quorum queue, if it is already declared as default, it will delete the existing one and create new HA queue
+        /// </summary>
+        /// <param name="cfg"></param>
+        /// <param name="configuration">Configuration</param>
+        public static void AddAsHighAvailableQueue(this IRabbitMqReceiveEndpointConfigurator cfg,
+                                      IConfiguration configuration)
+        {
+            cfg.SetQuorumQueue(3);
```
Comment (616e696c) [inline Carbon.MassTransit/IServiceCollectionConfiguratorExtensions.cs:293]: Can quorum count/value comes from settings?

File: Carbon.MassTransit/ReceiveEndpointObserver.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+                    var factory = new RabbitMQ.Client.ConnectionFactory()
+                    {
+                        Uri = new Uri(rabbitMqUql)
+                    };
+                    var absolutePath = faulted.ReceiveEndpoint.InputAddress.AbsolutePath.Split('/');
+                    var queueName = absolutePath[absolutePath.Length - 1];
+                    using (var connection = factory.CreateConnection())
+                    {
+                        using (var channel = connection.CreateModel())
+                            channel.QueueDelete(queueName, false, false);
```
Comment (616e696c) [inline Carbon.MassTransit/ReceiveEndpointObserver.cs:45]: Whats gonna happen if the queue is not empty? Isn't there a way to preserve messages?

Comment (keremaytac) [reply Carbon.MassTransit/ReceiveEndpointObserver.cs:45]: If it is not empty, queue will be purged and deleted. There is no way to wait until all the messages consumed and then delete. So they should handle with care while migrating.

File: Carbon.MassTransit/IServiceCollectionConfiguratorExtensions.cs (hunk 2 above)
Comment (keremaytac) [reply Carbon.MassTransit/IServiceCollectionConfiguratorExtensions.cs:293]: Actually, more than 5 replication of quorum is not recommended. As we have mostly 3 or 5 nodes of rabbit, and 3 is also a good way of quoruming, I kept it constant as 3.

1 is not quorum
2 is not recommended as it is not odd number
3 is OK
4 is not recommended as it is not odd number
5 is OK (has more burden comparing it to 3)
5+ not recommended

So, 3  always works best for us, also preventing it to be misconfigured by anyone else from configuration.

**Review by anil-eladag — APPROVED** [review]

**Pull request merged**